### PR TITLE
adding docs for compiler defined global variables

### LINKED ISSUE #15527: Add documentation for compiler-defined variables
Variables(or symbols?) like `boundsChecking` are defined by the compiler and have certain values. Users could run into unexpected behavior when they try to use the name of the symbol. 
Also, a module author had better respect these variables when trying to accomplish some internal checking.
So I think it's important to document this.
https://github.com/chapel-lang/chapel/blob/4d7fee8d0070b8336ccd52a683b26c5b34aa7520/compiler/AST/type.cpp#L881-L899

### PR DESCRIPTION
**Pull request by ram-nad**
Adding documentation for global variables defined by the compiler.

Closes #15527

### CODE REVIEW

**Comment by ram-nad** [comment]
I am not sure about use of `_privatization` and `chpl_nodeID`. Please help me in understanding their usage (and, whether I should add them to documentation). For `chpl_overloadSetsChecks` @lydia-duncan Maam pointed out that it is something compiler has to deal with (and also I could not find its usage in any of the Chapel modules) so I am not adding it in the list. Please correct me if I am wrong.

**Comment by lydia-duncan** [comment]
Names that start with `_` or `chpl_` should not be used by users and are reserved for internal names, so I don't think they need to be documented.

File: doc/rst/technotes/globalvars.rst [hunk 1]
```diff
@@ -0,0 +1,47 @@
+.. _readme-globalvars:
+
+==================================
+Global Variables
+==================================
+
+.. warning:: The variables defined here are not part of Chapel Standard and 
+             they can be changed, renamed or removed at any point in future.
+
+Chapel Compiler define some variables that are available to all the Chapel programs.
```
[the remark this answers is not in the document]
Comment (lydia-duncan) [reply doc/rst/technotes/globalvars.rst:10]: "define" -> "defines"

[the remark this answers is not in the document]
Comment (lydia-duncan) [reply doc/rst/technotes/globalvars.rst:10]: "all the Chapel programs" -> "all Chapel programs"

File: doc/rst/technotes/globalvars.rst [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+==================================
+Global Variables
+==================================
+
+.. warning:: The variables defined here are not part of Chapel Standard and 
+             they can be changed, renamed or removed at any point in future.
+
+Chapel Compiler define some variables that are available to all the Chapel programs. 
+These variables are used by compiler itself and standard Chapel libraries/packages
```
Comment (lydia-duncan) [inline doc/rst/technotes/globalvars.rst:11]: "by compiler" -> "by the compiler"

Comment (lydia-duncan) [reply doc/rst/technotes/globalvars.rst:11]: "and standard" -> "and the standard"

File: doc/rst/technotes/globalvars.rst [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+==================================
+Global Variables
+==================================
+
+.. warning:: The variables defined here are not part of Chapel Standard and 
+             they can be changed, renamed or removed at any point in future.
+
+Chapel Compiler define some variables that are available to all the Chapel programs. 
+These variables are used by compiler itself and standard Chapel libraries/packages 
+to provide some features/optimizations on the basis of compiler options.
```
Comment (lydia-duncan) [inline doc/rst/technotes/globalvars.rst:12]: "compiler options" -> "compiler flags"?  "compilation flags"?

File: doc/rst/technotes/globalvars.rst [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+Global Variables
+==================================
+
+.. warning:: The variables defined here are not part of Chapel Standard and 
+             they can be changed, renamed or removed at any point in future.
+
+Chapel Compiler define some variables that are available to all the Chapel programs. 
+These variables are used by compiler itself and standard Chapel libraries/packages 
+to provide some features/optimizations on the basis of compiler options. 
+All these variables are compile-time constants.
```
Comment (lydia-duncan) [inline doc/rst/technotes/globalvars.rst:13]: "All these" -> "All of these"

File: doc/rst/technotes/globalvars.rst [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+Variables
+---------
+
+**boundsChecking**
+
+    This variable can be used to check if ``bounds-checks`` is enabled or not.
```
Comment (lydia-duncan) [inline doc/rst/technotes/globalvars.rst:20]: "is enabled" -> "are enabled"

File: doc/rst/technotes/globalvars.rst [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+
+**boundsChecking**
+
+    This variable can be used to check if ``bounds-checks`` is enabled or not. 
+    It is used by libraries to enable/disable runtime bound checks while 
+    accessing ranges, arrays or domains.
+
+**castChecking**
+
+    This variable reflects wether ``cast-checks`` is enabled or not.
```
Comment (lydia-duncan) [inline doc/rst/technotes/globalvars.rst:26]: "wether" -> "whether"

Comment (lydia-duncan) [reply doc/rst/technotes/globalvars.rst:26]: "is enabled" -> "are enabled"

File: doc/rst/technotes/globalvars.rst [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+
+**chpl_checkNilDereferences**
+
+    When ``nil-checks`` is enabled, this variable is set to *true*. It is used to check if a object
+    is nil before accessing it.
+
+**chpl_checkDivByZero**
+
+    This variable can be used to check if ``div-by-zero-checks`` is enabled. When this variable is
+    set to *true*, the runtime will check for attempt to divide by zero.
```
Comment (lydia-duncan) [inline doc/rst/technotes/globalvars.rst:37]: "attempt" -> "attempts"

File: doc/rst/technotes/globalvars.rst [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+
+**chpl_warnUnstable**
+
+    This variable can be used to check if, ``warn-unstable`` compiler flag is enabled. When set to *true*,
```
Comment (lydia-duncan) [inline doc/rst/technotes/globalvars.rst:41]: Stray comma here, should be "check if ``warn-unstable``" instead

File: doc/rst/technotes/globalvars.rst [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+    compiler will issue warning on the use of code that can be changed in future or are not suitable for use.
```
Comment (lydia-duncan) [inline doc/rst/technotes/globalvars.rst:42]: "compiler" -> "the compiler"

Comment (lydia-duncan) [reply doc/rst/technotes/globalvars.rst:42]: "issue warning" -> "issue a warning"

Comment (lydia-duncan) [reply doc/rst/technotes/globalvars.rst:42]: "on the use of code" -> "about code"

Comment (lydia-duncan) [reply doc/rst/technotes/globalvars.rst:42]: "are not suitable" -> "is not suitable"

File: doc/rst/technotes/globalvars.rst [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+
+**_local**
+
+    This variable can be used to check if program is being compiled for singlw/multi locale excecution.
```
Comment (lydia-duncan) [inline doc/rst/technotes/globalvars.rst:46]: "if program" -> "if a program"

Comment (lydia-duncan) [reply doc/rst/technotes/globalvars.rst:46]: "singlw" -> "single"

File: doc/rst/technotes/globalvars.rst [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+    It is set to *true* if program is compiled for single locale excecution.
```
Comment (lydia-duncan) [inline doc/rst/technotes/globalvars.rst:47]: "if program" -> "if the program"

**Comment by ram-nad** [comment]
I am sorry about the silly grammatical mistakes. I have corrected all of them as suggested here. Please let me know if something is to be added or removed. Thanks

File: doc/rst/technotes/globalvars.rst [hunk 12]
```diff
@@ -0,0 +1,47 @@
+.. _readme-globalvars:
+
+==================================
+Global Variables
+==================================
+
+.. warning:: The variables defined here are not part of Chapel Standard and
```
Comment (cassella) [inline doc/rst/technotes/globalvars.rst:7]: Should this be "the Chapel Standard"?

File: doc/rst/technotes/globalvars.rst [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+These variables are used by the compiler itself and the standard Chapel libraries/packages 
+to provide some features/optimizations based on compilation flags. 
+All of these variables are compile-time constants.
+
+Variables
+---------
+
+**boundsChecking**
+
+    This variable can be used to check if ``bounds-checks`` are enabled or not.
```
Comment (cassella) [inline doc/rst/technotes/globalvars.rst:20]: I'd either refer to the compiler commandline option with dashes `--bounds-checks` as singular, or refer to "bounds checks" as plural.  This variable is about the commandline option, not general bounds checking, so I'd say it should be "... can be used to check if ``--bounds-checks`` is enabled or not."  or "if the bounds checking controlled by ``--bounds-checks`` are enabled or not.

Or even strike the "to check if", and say "This variable is true if ...", and drop the "or not" at the end.

Similarly for the other `--*-checks` command-line options.

File: doc/rst/technotes/globalvars.rst [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+**boundsChecking**
+
+    This variable can be used to check if ``bounds-checks`` are enabled or not. 
+    It is used by libraries to enable/disable runtime bound checks while 
+    accessing ranges, arrays, or domains.
+
+**castChecking**
+
+    This variable reflects whether ``cast-checks`` are enabled or not.
+    It is used to provide checks before casting a variable from one type to another.
```
Comment (cassella) [inline doc/rst/technotes/globalvars.rst:27]: In reviewing the use of this variable, it seems to apply in very limited circumstances:

`integral.safeCast()`, and `channel`'s `readbits()` and `writebits()` methods(?).  So it doesn't apply to most casts.

File: doc/rst/technotes/globalvars.rst [hunk 15]
```diff
@@ -0,0 +1,47 @@
+.. _readme-globalvars:
+
+==================================
+Global Variables
```
Comment (cassella) [inline doc/rst/technotes/globalvars.rst:4]: I think Brad objects to the phrase "global variables", but I don't know what else to call them.  (Or if these are actually outside the scope of the outermost module.)  And to nitpick, I think they're all params, not variables.  :)

File: doc/rst/technotes/globalvars.rst [hunk 16]
```diff
@@ -0,0 +1,47 @@
...
+to provide some features/optimizations based on compilation flags. 
+All of these variables are compile-time constants.
+
+Variables
+---------
+
+**boundsChecking**
+
+    This variable can be used to check if ``bounds-checks`` are enabled or not. 
+    It is used by libraries to enable/disable runtime bound checks while
```
Comment (cassella) [inline doc/rst/technotes/globalvars.rst:21]: I would "bound" -> "bounds" here.

File: doc/rst/technotes/globalvars.rst (hunk 15 above)
Comment (ram-nad) [reply doc/rst/technotes/globalvars.rst:4]: Should it be Global Params then?

Comment (lydia-duncan) [reply doc/rst/technotes/globalvars.rst:4]: I think Brad mostly objects to calling symbols defined at module scope "global".  I think these could reasonably be called "global", though.

File: doc/rst/technotes/globalvars.rst [hunk 17]
```diff
@@ -0,0 +1,47 @@
...
+
+**boundsChecking**
+
+    This variable is *true* when ``--bounds-checks`` is enabled. 
+    It is used by libraries to enable/disable runtime bounds checks while 
+    accessing ranges, arrays, or domains.
+
+**castChecking**
+
+    This variable reflects whether ``--cast-checks`` flag is enabled.
```
Comment (lydia-duncan) [inline doc/rst/technotes/globalvars.rst:26]: I think it should be "the ``--cast-checks`` flag"

File: doc/rst/technotes/globalvars.rst (hunk 15 above)
Comment (ram-nad) [reply doc/rst/technotes/globalvars.rst:4]: Maam is the name 'Global Variables` fine then?

File: doc/rst/technotes/globalvars.rst (hunk 14 above)
Comment (ram-nad) [reply doc/rst/technotes/globalvars.rst]: All these casts apply to situation where there is a chance of losing information or narrowing values, should this be mentioned explicitly?

File: doc/rst/technotes/globalvars.rst (hunk 17 above)
Comment (ram-nad) [reply doc/rst/technotes/globalvars.rst:26]: Sure Maam I will update it.

File: doc/rst/technotes/globalvars.rst (hunk 14 above)
Comment (lydia-duncan) [reply doc/rst/technotes/globalvars.rst]: > All these casts apply to situation where there is a chance of losing information or narrowing values, should this be mentioned explicitly?

That seems reasonable to me.

> In reviewing the use of this variable, it seems to apply in very limited circumstances

I think "in some cases" would be enough wiggle words to indicate that it isn't necessarily applied everywhere that it could be.  It also means that its use can be expanded without requiring us to be constantly updating this documentation

File: doc/rst/technotes/globalvars.rst (hunk 15 above)
Comment (lydia-duncan) [reply doc/rst/technotes/globalvars.rst:4]: I think it's fine.  Variables can be specifically those that are declared `var`, or they can also be the general category of symbols which includes `var`, `param`, etc.  At least, that's the impression I'm getting from the spec's [section on the subject](https://chapel-lang.org/docs/latest/language/spec/variables.html)

**Comment by ram-nad** [comment]
> Looks good, and the built docs also look good. Thanks!

Thank you, Ma'am.
